src/library/coverartdelegate.cpp

```diff
@@ -71,25 +77,41 @@ void CoverArtDelegate::slotOnlyCachedCoverArt(bool b) {
     }
 }
 
-void CoverArtDelegate::slotCoverFound(const QObject* pRequestor,
-                                      const CoverInfoRelative& info,
-                                      QPixmap pixmap, bool fromCache) {
-    if (pRequestor == this && !pixmap.isNull() && !fromCache) {
-        // qDebug() << "CoverArtDelegate::slotCoverFound" << pRequestor << info
-        //          << pixmap.size();
-        QLinkedList<int> rows = m_hashToRow.take(info.hash);
-        foreach(int row, rows) {
-            emit coverReadyForCell(row, m_iCoverColumn);
+void CoverArtDelegate::slotCoverFound(
+        const QObject* pRequestor,
+        const CoverInfo& coverInfo,
+        const QPixmap& pixmap,
+        quint16 requestedHash,
+        bool coverInfoUpdated) {
+    Q_UNUSED(pixmap);
+    if (pRequestor != this) {
+        return;
+    }
+    const QLinkedList<int> rows =
+            m_hashToRow.take(requestedHash);
+    foreach(int row, rows) {
+        emit coverReadyForCell(row, m_iCoverColumn);
+    }
+    if (m_pTrackModel && coverInfoUpdated) {
+        const auto pTrack =
+                m_pTrackModel->getTrackByRef(
+                        TrackRef::fromFileInfo(coverInfo.trackLocation));
+        if (pTrack) {
+            kLogger.info()
+                    << "Updating cover info of track"
+                    << coverInfo.trackLocation;
+            pTrack->setCoverInfo(coverInfo);
```

Comment:
This is redundant if we pass the track into the request. 
Can't we do it at one place only?

Reply:
No, we don't. The CoverArtDelegate accesses query results directly. Track objects are only loaded if the CoverArtCache signals that the cover info has been updated.